lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -0,0 +1,162 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Cardano.Wallet.DB.Sqlite where
+
+import Prelude
+
+import Cardano.Wallet.DB.SqliteTypes
+    ( AddressScheme, TxId )
+import Data.Text
+    ( Text )
+import Data.Time.Clock
+    ( UTCTime )
+import Data.Word
+    ( Word32 )
+import Database.Persist.TH
+import GHC.Generics
+    ( Generic (..) )
+
+import qualified Cardano.Wallet.Primitive.Types as W
+
+-- fixme: need tables for wallet AddressPool
+
+share
+    [ mkPersist sqlSettings { mpsPrefixFields = False }
+    , mkMigrate "migrateAll"
+    ]
+    [persistLowerCase|
+
+-- Wallet IDs, address discovery state, and metadata.
+Wallet
+    walId                       W.WalletId     sql=wallet_id
+    walName                     Text           sql=name
+    walPassphraseLastUpdatedAt  UTCTime        sql=passphrase_last_updated_at
+    walStatus                   W.WalletState  sql=status
+    walDelegation               Text Maybe     sql=delegation
+    walAddressScheme            AddressScheme  sql=address_discovery
+
+    Primary walId
+    deriving Show Generic
+
+-- The private key for each wallet. This is in a separate table simply so that
+-- "SELECT * FROM wallet" won't print keys.
+WalletPrivateKey                    sql=private_key
+    walPrivateKeyWalId  W.WalletId  sql=wallet_id
+    walPrivateKey       Text        sql=private_key
+
+    Primary walPrivateKeyWalId
+    Foreign Wallet fk_wallet_private_key walPrivateKeyWalId
+
+    deriving Show Generic
+
+-- Maps a transaction ID to its metadata (which is calculated when applying
+-- blocks).
+--
+-- TxMeta is specific to a wallet because multiple wallets may have the same
+-- transaction with different metdata values. The associated inputs and outputs
+-- of the transaction are in the TxIn and TxOut tables.
+TxMeta
+    txId             TxId         sql=tx_id
+    txMetaWalletId   W.WalletId   sql=wallet_id
+    txStatus         W.TxStatus   sql=status
+    txMetaDirection  W.Direction  sql=direction
+    txMetaSlotId     W.SlotId     sql=slot_id
+    txMetaAmount     W.Coin       sql=amount
+
+    Primary txId txMetaWalletId
+    Foreign Wallet fk_wallet_tx_meta txMetaWalletId
+    deriving Show Generic
+
+-- A transaction input associated with TxMeta.
+--
+-- There is no wallet ID because these values depend only on the transaction,
+-- not the wallet. txInputTxId is referred to by TxMeta and PendingTx
+TxIn
+    txInputTxId         TxId    sql=tx_id
+    txInputSourceTxId   TxId    sql=source_id
+    txInputSourceIndex  Word32  sql=source_index
+    txInputAddress      Text    sql=address
+    txInputAmount       W.Coin  sql=amount
```

Comment:
Note: we won't have the input address and amount here. That's some something that can be trivially retrieved and, we don't need to actually store it.

Reply:
Also, we do store the output index but not the input index in the tx? We do need to preserve the order of inputs when retrieving a transaction from the DB (as the binary representation depends on it!), so we also need to precisely identify the input with their index in the transaction.

Reply:
removed `txInputAddress` and `txInputAmount` . TxIns have already index. Added *Table* everywhere in the generated fields as Matt Parsons used to do. Still, I feel the schema part needs **due dilligence**